src/python/pants/engine/internals/selectors.py

```diff
@@ -701,6 +710,9 @@ def __init__(self, *args: Any) -> None:
 def native_engine_generator_send(
     func, arg
 ) -> PyGeneratorResponseGet | PyGeneratorResponseGetMulti | PyGeneratorResponseBreak:
+    if _ENABLE_DEBUGPY:
+        debugpy.debug_this_thread()
```

Comment:
The engine only uses a few dozen threads, and it will reuse them repeatedly. Probably rather than enabling it here, you'd want to enable it via https://docs.rs/tokio/latest/tokio/runtime/struct.Builder.html#method.on_thread_start here: https://github.com/pantsbuild/pants/blob/21b14c5971c41f67217027cc431611ba4448a59f/src/rust/engine/task_executor/src/lib.rs#L81-L86

Reply:
Hmmmm for this to work, we'd need to access the python runtime already, right? I wonder if its worth the headache because we're already debugging so performance isn't that much of an issue...

Reply:
We still end up creating the executor explicitly from the options:
```
$ rg 'PyExecutor\(' src/python/pants/
src/python/pants/engine/rules_test.py
36:_EXECUTOR = PyExecutor(core_threads=2, max_threads=4)

src/python/pants/engine/internals/scheduler_test_base.py
22:    _executor = PyExecutor(core_threads=2, max_threads=4)

src/python/pants/testutil/rule_runner.py
157:_EXECUTOR = PyExecutor(core_threads=1, max_threads=3)

src/python/pants/option/global_options.py
1808:        return PyExecutor(
```

Reply:
...oh. Hm. This is set via an environment variable though? That probably won't work if we want to set it as this level. Would need to be a bootstrap option.

Reply:
The bootstrap option could be defaulted from env var though, right?

My aim here is to also add support to the user-facing pants bootstrap script so plugin authors can debug their rules.

Reply:
Hmmm, I can't hit my breakpoints with the following:

```rust
    let mut runtime_builder = Builder::new_multi_thread();

    runtime_builder.worker_threads(num_worker_threads)
      .max_blocking_threads(max_threads - num_worker_threads)
      .enable_all();

    if env::var("PANTS_DEBUG").is_ok() {
      runtime_builder.on_thread_start(|| {
        Python::with_gil(|py| {
          let _ = py.eval("__import__('debugpy').debug_this_thread()", None, None);
        } )
      });
    }

    let runtime = runtime_builder
      .build()
      .map_err(|e| format!("Failed to start the runtime: {}", e))?;
```

I know for a fact that code isn't returning an `Err`.

Reply:
I'm also sure the thread the rule I'm debugging is running on has had this code executed (by some print-debugging)